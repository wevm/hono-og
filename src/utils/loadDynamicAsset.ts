import type { SatoriOptions } from "satori";
import { apis, getIconCode, loadEmoji } from "./emoji.js";
import { loadGoogleFont } from "./loadGoogleFont.js";

const languageFontMap = {
  "ja-JP": "Noto Sans JP",
  "ko-KR": "Noto Sans KR",
  "zh-CN": "Noto Sans SC",
  "zh-TW": "Noto Sans TC",
  "zh-HK": "Noto Sans HK",
  "th-TH": "Noto Sans Thai",
  "bn-IN": "Noto Sans Bengali",
  "ar-AR": "Noto Sans Arabic",
  "ta-IN": "Noto Sans Tamil",
  "ml-IN": "Noto Sans Malayalam",
  "he-IL": "Noto Sans Hebrew",
  "te-IN": "Noto Sans Telugu",
  devanagari: "Noto Sans Devanagari",
  kannada: "Noto Sans Kannada",
  symbol: ["Noto Sans Symbols", "Noto Sans Symbols 2"],
  math: "Noto Sans Math",
  unknown: "Noto Sans",
};

export async function loadDynamicAsset(
  emojiType: keyof typeof apis,
  code: string,
  text: string,
): Promise<string | SatoriOptions["fonts"]> {
  if (code === "emoji")
    // It's an emoji, load the image.
    return `data:image/svg+xml;base64,${btoa(
      await loadEmoji(emojiType, getIconCode(text)),
    )}`;

  const codes = code.split("|");

  // Try to load from Google Fonts.
  const names = codes
    .map((code) => languageFontMap[code as keyof typeof languageFontMap])
    .filter(Boolean);

  try {
    return await Promise.all(
      names.flat().map(async (name, index) => {
        return {
          name: `satori_${codes[index]}_fallback_${text}`,
          data: await loadGoogleFont({ family: name, weight: 400 }),
          weight: 400 as const,
          style: "normal" as const,
          lang: codes[index] as string,
        };
      }),
    );
  } catch (e) {
    console.error("Failed to load dynamic font for", text, ". Error:", e);
    return [];
  }
}
