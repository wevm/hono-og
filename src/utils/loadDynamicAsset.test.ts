import { expect, test } from "vitest";
import { loadDynamicAsset } from "./loadDynamicAsset.js";

test("emoji", async () => {
  expect(await loadDynamicAsset("twemoji", "emoji", "ðŸ‘‹")).toMatchSnapshot();
});

test("lang", async () => {
  expect(
    await loadDynamicAsset("twemoji", "ja-JP|zh-CN|zh-TW|zh-HK", "ä½ å¥½"),
  ).toMatchSnapshot();
});
