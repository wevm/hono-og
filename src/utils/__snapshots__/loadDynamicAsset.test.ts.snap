// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`emoji 1`] = `"data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCAzNiAzNiI+PHBhdGggZmlsbD0iI0VGOTY0NSIgZD0iTTQuODYxIDkuMTQ3Yy45NC0uNjU3IDIuMzU3LS41MzEgMy4yMDEuMTY2bC0uOTY4LTEuNDA3Yy0uNzc5LTEuMTExLS41LTIuMzEzLjYxMi0zLjA5MyAxLjExMi0uNzc3IDQuMjYzIDEuMzEyIDQuMjYzIDEuMzEyLS43ODYtMS4xMjItLjYzOS0yLjU0NC40ODMtMy4zMzEgMS4xMjItLjc4NCAyLjY3LS41MTMgMy40NTYuNjExbDEwLjQyIDE0LjcyTDI1IDMxbC0xMS4wODMtNC4wNDJMNC4yNSAxMi42MjVjLS43OTMtMS4xMjktLjUxOS0yLjY4Ni42MTEtMy40Nzh6Ii8+PHBhdGggZmlsbD0iI0ZGREM1RCIgZD0iTTIuNjk1IDE3LjMzNnMtMS4xMzItMS42NS41MTktMi43ODFjMS42NDktMS4xMzEgMi43OC41MTggMi43OC41MThsNS4yNTEgNy42NThjLjE4MS0uMzAyLjM3OS0uNi42LS44OTRMNC41NTcgMTEuMjFzLTEuMTMxLTEuNjQ5LjUxOS0yLjc4YzEuNjQ5LTEuMTMxIDIuNzguNTE4IDIuNzguNTE4bDYuODU1IDkuOTk3Yy4yNTUtLjIwOC41MTYtLjQxNy43ODUtLjYyMkw3LjU0OSA2LjczMnMtMS4xMzEtMS42NDkuNTE5LTIuNzhjMS42NDktMS4xMzEgMi43OC41MTggMi43OC41MThsNy45NDcgMTEuNTg5Yy4yOTItLjE3OS41ODEtLjMzNC44NzEtLjQ5OEwxMi4yMzggNC43MjlzLTEuMTMxLTEuNjQ5LjUxOC0yLjc4YzEuNjQ5LTEuMTMxIDIuNzguNTE4IDIuNzguNTE4bDcuODU0IDExLjQ1NCAxLjE5NCAxLjc0MmMtNC45NDggMy4zOTQtNS40MTkgOS43NzktMi41OTIgMTMuOTAyLjU2NS44MjUgMS4zOS4yNiAxLjM5LjI2LTMuMzkzLTQuOTQ5LTIuMzU3LTEwLjUxIDIuNTkyLTEzLjkwM0wyNC41MTUgOC42MnMtLjU0NS0xLjkyNCAxLjM3OC0yLjQ3YzEuOTI0LS41NDUgMi40NyAxLjM3OSAyLjQ3IDEuMzc5bDEuNjg1IDUuMDA0Yy42NjggMS45ODQgMS4zNzkgMy45NjEgMi4zMiA1LjgzMSAyLjY1NyA1LjI4IDEuMDcgMTEuODQyLTMuOTQgMTUuMjc5LTUuNDY1IDMuNzQ3LTEyLjkzNiAyLjM1NC0xNi42ODQtMy4xMUwyLjY5NSAxNy4zMzZ6Ii8+PGcgZmlsbD0iIzVEQURFQyI+PHBhdGggZD0iTTEyIDMyLjA0MkM4IDMyLjA0MiAzLjk1OCAyOCAzLjk1OCAyNGMwLS41NTMtLjQwNS0xLS45NTgtMXMtMS4wNDIuNDQ3LTEuMDQyIDFDMS45NTggMzAgNiAzNC4wNDIgMTIgMzQuMDQyYy41NTMgMCAxLS40ODkgMS0xLjA0MnMtLjQ0Ny0uOTU4LTEtLjk1OHoiLz48cGF0aCBkPSJNNyAzNGMtMyAwLTUtMi01LTUgMC0uNTUzLS40NDctMS0xLTFzLTEgLjQ0Ny0xIDFjMCA0IDMgNyA3IDcgLjU1MyAwIDEtLjQ0NyAxLTFzLS40NDctMS0xLTF6TTI0IDJjLS41NTIgMC0xIC40NDgtMSAxcy40NDggMSAxIDFjNCAwIDggMy41ODkgOCA4IDAgLjU1Mi40NDggMSAxIDFzMS0uNDQ4IDEtMWMwLTUuNTE0LTQtMTAtMTAtMTB6Ii8+PHBhdGggZD0iTTI5IC4wNDJjLS41NTIgMC0xIC40MDYtMSAuOTU4cy40NDggMS4wNDIgMSAxLjA0MmMzIDAgNC45NTggMi4yMjUgNC45NTggNC45NTggMCAuNTUyLjQ4OSAxIDEuMDQyIDFzLjk1OC0uNDQ4Ljk1OC0xQzM1Ljk1OCAzLjE2MyAzMyAuMDQyIDI5IC4wNDJ6Ii8+PC9nPjwvc3ZnPg=="`;

exports[`lang 1`] = `
[
  {
    "data": ArrayBuffer [],
    "lang": "ja-JP",
    "name": "satori_ja-JP_fallback_你好",
    "style": "normal",
    "weight": 400,
  },
  {
    "data": ArrayBuffer [],
    "lang": "zh-CN",
    "name": "satori_zh-CN_fallback_你好",
    "style": "normal",
    "weight": 400,
  },
  {
    "data": ArrayBuffer [],
    "lang": "zh-TW",
    "name": "satori_zh-TW_fallback_你好",
    "style": "normal",
    "weight": 400,
  },
  {
    "data": ArrayBuffer [],
    "lang": "zh-HK",
    "name": "satori_zh-HK_fallback_你好",
    "style": "normal",
    "weight": 400,
  },
]
`;
